dhall-docs/src/Dhall/Docs/CodeRenderer.hs

```diff
@@ -0,0 +1,183 @@
+    go (currLine, currCol) currentLines ((Src {..}, import_) : rest) = do
+        let sourceLine = SourcePos.unPos . SourcePos.sourceLine
+        let sourceColumn = SourcePos.unPos . SourcePos.sourceColumn
+
+        let importStartLine = sourceLine srcStart
+        let importEndLine = sourceLine srcEnd
+
+        let importStartCol = sourceColumn srcStart
+        let importEndCol = sourceColumn srcEnd
+
+        let (prefixLines, restLines) = splitAt (importStartLine - currLine) currentLines
+        let (importLines, suffixLines) = splitAt (importStartLine - importStartLine + 1) restLines
+
+        -- calls to `head` and `last` here should never fail since `importLines`
+        -- have at least one element
+        let (firstImportLine, lastImportLine) = (head importLines, last importLines)
+        let prefixCols = Data.Text.take (importStartCol - currCol) firstImportLine
+        let suffixCols = Data.Text.drop (importEndCol - currCol) lastImportLine
+
+        -- prefix lines and columns
+        mapM_ (\t -> toHtml t >> br_ []) prefixLines
+        toHtml prefixCols
+
+        -- rendered import
+        renderImport import_ srcText
+
+        -- add a newline if last line of import consumes the remaining line on
+        -- the original text
+        if Data.Text.null suffixCols then br_ [] else return ()
+
+        let suffix = if Data.Text.null suffixCols then suffixLines else suffixCols : suffixLines
+
+        -- move the cursor to next line if no characterse are remaining on the
+        -- suffix cols, otherwise keep the last line and next char right after
+        -- the import. This is done to handle properly several imports on the
+        -- same line
+        let nextPosition = if Data.Text.null suffixCols then
+                               (importEndLine + 1, 1)
+                           else (importEndLine, importEndCol)
+
+        go nextPosition suffix rest
+
+-- | Internal utility to differentiate if a Dhall expr is a type annotation
+--   or the whole file
+data ExprType = TypeAnnotation | AssertionExample
+
+-- | Renders an AST /fragment/ from the source file AST. Use this only for that
+--   purpose. The difference between this and 'exprSrcToHtml' is because that
+--   the extracted fragment's 'SourcePos's need to be re-generated to
+--   render them in a better way; just adding whitespace at the beginning of the
+--   first line won't render good results.
+exprVoidToHtml :: Dhall.Pretty.CharacterSet -> ExprType -> Expr Void Import -> Html ()
+exprVoidToHtml characterSet exprType expr = exprSrcToHtml formattedFile expr'
+  where
+    layout = case exprType of
+        AssertionExample -> Dhall.Pretty.layout
+        TypeAnnotation -> typeLayout
+
+    formattedFile = Pretty.Text.renderStrict
+        $ layout
+        $ Dhall.Pretty.prettyCharacterSet characterSet (Core.denote expr)
+
+    expr' = case Dhall.Parser.exprFromText "" formattedFile of
```

Comment:
Why is it necessary to parse the expression from the rendered text? Can't we use the original expression?

Reply:
because we new to recover the new `Note`s and `SourcePos`s